File: DAA/Assignment-3/FractionalKnapsack.java
import java.util.*;

class Item {
    int value, weight;

    public Item(int value, int weight) {
        this.value = value;
        this.weight = weight;
    }
}

public class FractionalKnapsack {

    public static double fractionalKnapsack(int capacity, Item[] items) {
        Arrays.sort(items, (a, b) -> ((b.value / b.weight) - (a.value / a.weight)));
        double totalValue = 0;
        for (Item item : items) {
            if (capacity >= item.weight) {
                totalValue += item.value;
                capacity -= item.weight;
            } else {
                totalValue += ((double) item.value / item.weight) * capacity;
                break;
            }
        }
        return totalValue;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the capacity of the knapsack: ");
        int capacity = scanner.nextInt();
        System.out.print("Enter the number of items: ");
        int n = scanner.nextInt();
        Item[] items = new Item[n];
        System.out.println("Enter the value and weight of each item:");
        for (int i = 0; i < n; i++) {
            System.out.print("Value of item " + (i + 1) + ": ");
            int value = scanner.nextInt();
            System.out.print("Weight of item " + (i + 1) + ": ");
            int weight = scanner.nextInt();
            items[i] = new Item(value, weight);
        }
        scanner.close();
        double totalValue = fractionalKnapsack(capacity, items);
        System.out.println("The maximum value that can be obtained is: " + totalValue);
    }
}

File: pythoncodes/python/Practical03/FractionalKnapsack.py
import time
start = time.time()

def fractional_knapsack(items, capacity):
    # Calculate the value-to-weight ratio for each item
    item_value_ratio = [(item[1] / item[0], item) for item in items]
    
    # Sort items by value-to-weight ratio in descending order
    item_value_ratio.sort(reverse=True)
    
    total_value = 0  # Total value of selected items
    knapsack = []    # Items selected for the knapsack
    
    for value_per_weight, item in item_value_ratio:
        if capacity >= item[0]:  # If the item can fit entirely
            knapsack.append((item, 1))  # Add the whole item
            total_value += item[1]
            capacity -= item[0]
        else:  # If the item can only fit partially
            fraction = capacity / item[0]
            knapsack.append((item, fraction))  # Add a fraction of the item
            total_value += item[1] * fraction
            break  # The knapsack is now full
    
    return total_value, knapsack

# Example usage:
items = [(2, 10), (3, 5), (5, 15), (7, 7), (1, 6)]
capacity = 10
max_value, selected_items = fractional_knapsack(items, capacity)

print("Maximum value: {}".format(max_value))
print("Selected items:")
for item, fraction in selected_items:
    print("  {}: Fraction = {}".format(item,round(fraction,2)))

end = time.time()
print("\nExecution time is: {}ms".format((end-start)*10**3))